src/hotspot/share/services/memReporter.cpp

```diff
@@ -820,3 +822,68 @@ void MemDetailDiffReporter::diff_virtual_memory_site(const NativeCallStack* stac
 
   out->print_cr(")\n");
  }
+
+void MemJFRReporter::sendTotalEvent() {
+  if (!MemTracker::enabled()) {
+    return;
+  }
+
+  MemBaseline usage;
+  usage.baseline(true);
+
+  const size_t malloced_memory = usage.malloc_memory_snapshot()->total();
+  const size_t reserved_memory = usage.virtual_memory_snapshot()->total_reserved();
+  const size_t committed_memory = usage.virtual_memory_snapshot()->total_committed();
+
+  const size_t reserved = malloced_memory + reserved_memory;
+  const size_t committed = malloced_memory + committed_memory;
+
+  EventNativeMemoryUsage event;
+  event.set_reserved(reserved);
+  event.set_committed(committed);
+  event.commit();
+}
+
+void MemJFRReporter::sendTypeEvent(const char* type, size_t reserved, size_t committed) {
+  EventNativeMemoryUsagePart event;
+  event.set_type(type);
+  event.set_reserved(reserved);
+  event.set_committed(committed);
+  event.commit();
+}
+
+void MemJFRReporter::sendTypeEvents() {
+  if (!MemTracker::enabled()) {
+    return;
+  }
+
+  MemBaseline usage;
+  usage.baseline(true);
```

Comment:
Note that this is quite expensive. So it depends on how often we do this. How often are these samples taken?

Eg. Baseline.baseline does walk all thread stacks (so, probing all of them via mincore()). It also copies a lot of counters around. 

It is also not threadsafe. Are we at a safepoint here? Normally NMT reports are only done at safepoints.

Reply:
Haven't looked at the details of `baseline(bool summaryOnly)` that much, but since it `summaryOnly = true` I don't think it actually walk the thread stacks, right? 

We don't do this at a safepoint but looking at `MallocMemorySnapshot::copy_to(...)` it uses `ThreadCritical` to avoid things being cleaned out at the same time. Not sure if there are other thread safety problems though. I would expect this to have the same problems as a summary report triggered through `jcmd` because that isn't run at a safepoint either. But I do see that when used from `jcmd` we take a lock to serialize the NMT query, so we should probably do the same here.
```
  // Query lock is used to synchronize the access to tracking data.
  // So far, it is only used by JCmd query, but it may be used by
  // other tools.
  static inline Mutex* query_lock() {
    assert(NMTQuery_lock != NULL, "not initialized!");
    return NMTQuery_lock;
  }
```
So we certainly need to look closer at this. I would like to understand why the query lock is needed.

Reply:
Looking a bit more at this. To me it seems like we are protecting the stored baseline with the lock is:
```
class MemTracker : AllStatic {
...
  // Stored baseline
  static MemBaseline      _baseline;
...
};
```
That the `NMTDCmd` code uses through `MemTracker::get_baseline()`. So if the JFR events keep its own baseline we should be fine without the lock. Or am I missing something here?

Reply:
Hi @kstefanj ,

> Haven't looked at the details of `baseline(bool summaryOnly)` that much, but since it `summaryOnly = true` I don't think it actually walk the thread stacks, right?

I think we do, since summary info contains information about committed thread stack size. I think the call chain is 

```
MemBaseline::baseline()->
MemBaseline::baseline_summary()->
VirtualMemorySummary::snapshot()->
VirtualMemoryTracker::snapshot_thread_stacks()->
VirtualMemoryTracker::walk_virtual_memory()->
SnapshotThreadStackWalker::do_allocation_site()->
RegionIterator::next_committed()->
os::committed_in_range()->
mincore loop
```

> 
> We don't do this at a safepoint but looking at `MallocMemorySnapshot::copy_to(...)` it uses `ThreadCritical` to avoid things being cleaned out at the same time. Not sure if there are other thread safety problems though. I would expect this to have the same problems as a summary report triggered through `jcmd` because that isn't run at a safepoint either.

You are right, I could have sworn we do this with a VM op. Must have confused this with metaspace printing.

> But I do see that when used from `jcmd` we take a lock to serialize the NMT query, so we should probably do the same here.
> 
> ```
>   // Query lock is used to synchronize the access to tracking data.
>   // So far, it is only used by JCmd query, but it may be used by
>   // other tools.
>   static inline Mutex* query_lock() {
>     assert(NMTQuery_lock != NULL, "not initialized!");
>     return NMTQuery_lock;
>   }
> ```
> 
> So we certainly need to look closer at this. I would like to understand why the query lock is needed.

To serialize jcmds running in parallel. Since jcmd is done by the attach listener thread, the only way I can think of would be if someone in parallel executes the command via mbean? Not sure.

Note that there is no way to get consistent counter values across all categories anyway. All the counters are updated atomically without lock, and that is all we can afford in malloc. Counters are not synchronized among themselves.

Reply:
> I think we do, since summary info contains information about committed thread stack size. 

You are totally correct, somehow missed that part of `VirtualMemorySummary::snapshot()`. So yes, we should really look at only doing one snapshot per period. Any feeling for exactly how expensive this walk is?

> To serialize jcmds running in parallel. Since jcmd is done by the attach listener thread, the only way I can think of would be if someone in parallel executes the command via mbean? Not sure.

Ok, so if executed via the mbean it would be executed in another thread but still use `DCMD::execute()` and the lock to make sure those are synchronized. Anyhow, do you agree with my above understanding that what is protected by the lock is the "cached baseline"? So as long as JFR doesn't use it, the lock should not be needed?

> Note that there is no way to get consistent counter values across all categories anyway. All the counters are updated atomically without lock, and that is all we can afford in malloc. Counters are not synchronized among themselves.

Agreed, this is a snapshot at a given point of a moving target.

Reply:
Yes, maybe creating a special baseline-call that skips the thread-stacks would be ok. As you say, if these are expensive it might be problematic to do them in the periodic thread.